spin_coater.c

```diff
@@ -0,0 +1,196 @@
+    return false;
+
+  ctx->spin_coater.pwm_duty = duty;
+  DEBUG_printf("\n PWM set to: %lu!\n", duty);
+  return pwm_set_freq_duty(ctx->spin_coater.pwm_slice_num, PWM_CHAN_A, PWM_490_FREQ, duty);
+}
+#endif
+
+int64_t
+timer_spin_callback(alarm_id_t id, void* user_data)
+{
+  furnace_context_t *ctx = (furnace_context_t *)user_data;
+
+  ctx->spin_coater.spin_state = SPIN_SMOOTH_STOP_REQUESTED;
+  const char* msg = "spin_stopped\n";
+  tcp_server_send_data(
+    ctx, ctx->tcp.client_pcb, (const uint8_t*)msg, strlen(msg));
+
+  return 0;
+}
+
+#if CONFIG_SPIN_COATER_DSHOT
+int scale_dshot_value_when_speeding(furnace_context_t* ctx, int rpm_left, int bias){
+    return (50.0 / ctx->spin_coater.set_rpm) * rpm_left + bias;
+}
+
+int scale_dshot_value_when_stopping(furnace_context_t* ctx, int rpm_left){
+    return 50.0 - (50.0 / ctx->spin_coater.set_rpm) * rpm_left;
+}
+
+absolute_time_t do_dshot_smooth_transition(furnace_context_t* ctx)
+{
+  int direction = current_rpm < ctx->spin_coater.set_rpm ? 1 : -1;
+  int rpm_diff = ctx->spin_coater.set_rpm - current_rpm;
+  absolute_time_t next_delay;
+  if(SPIN_STARTED_WITH_TIMER == ctx->spin_coater.spin_state){
+    if(abs(rpm_diff) > 100)
+      ctx->spin_coater.dshot_throttle_val += scale_dshot_value_when_speeding(ctx,
+                                                                            abs(rpm_diff),
+                                                                            10*(current_rpm/(double)(ctx->spin_coater.set_rpm)))*direction;
+    else if (abs(rpm_diff) > 10 )
+      ctx->spin_coater.dshot_throttle_val += 1*direction;
+    next_delay = make_timeout_time_ms(ctx->spin_coater.rpm_speedup_update_delay);
+  }
+  else if(SPIN_SMOOTH_STOP_REQUESTED == ctx->spin_coater.spin_state) {
+    ctx->spin_coater.dshot_throttle_val -= scale_dshot_value_when_stopping(ctx, abs(rpm_diff));
+      if(abs(rpm_diff) > (ctx->spin_coater.set_rpm - 250) || ctx->spin_coater.dshot_throttle_val < MIN_THROTTLE_COMMAND) {
+        ctx->spin_coater.dshot_throttle_val = MIN_THROTTLE_COMMAND;
+        ctx->spin_coater.spin_state = SPIN_IDLE;
+      }
+    next_delay = make_timeout_time_ms(ctx->spin_coater.rpm_slowdown_update_delay);
+  }else {
+   next_delay = make_timeout_time_ms(1000); 
+  }
+  set_dshot_safe(ctx, ctx->spin_coater.dshot_throttle_val);
+  return next_delay;
+}
+#elif CONFIG_SPIN_COATER_PWM
+absolute_time_t do_pwm_smooth_transition(furnace_context_t* ctx)
+{
+  int direction = current_rpm < ctx->spin_coater.set_rpm ? 1 : -1;
+
+  ctx->spin_coater.pwm_duty += direction;
+  set_spin_coater_pwm_safe(ctx, ctx->spin_coater.pwm_duty);
+
+  return make_timeout_time_ms(100);
+}
+#endif
+
+static void
+do_spin_coater_rpm_log_update(furnace_context_t *ctx, bool deadline_met)
+{
+  const char* rpm_str = "rpm: %u\n";
+  char send_buf[BUF_SIZE];
+
+  if (deadline_met && ctx->spin_coater.spin_state != SPIN_IDLE) {
+    int number_of_chars =
+      snprintf(send_buf,
+               BUF_SIZE,
```

Comment:
`sizeof(send_buf)` or `ARRAY_SIZE(send_buf)`?  
To be fair, we should be using `ARRAY_SIZE()` elsewhere in the code too..

Reply:
right now I changed that to `sizeof(send_buf)`, probably we can refactor whole code at once to sub with `ARRAY_SIZE(send_buf)`